Added ls command

CollCaz opened the pull request:
Added the ls command proposed by @tryoxiss in issue #6 .

With no arguments; the command will print the tree starting at the root of the current scene.
You can specify an absolute path to any node in the current scene or an absolute path to any .tscn file.

I might revisit this command in the future to clean up the code and possible add more features, but for now i am sick of it.

edit: you don't have to type /root it will do that for you if you don't.

CollCaz committed:
Added ls command

Kubulambula commented:
Thank you!
I'll check it out and review it as soon as I get home.

CollCaz commented:
Let me know if anything needs fixing!

Kubulambula commented on addons/gdshell/commands/default_commands/ls.gd:
```diff
@@ -0,0 +1,73 @@
+
+func output_tree(path) -> Dictionary:
+	var node: Node
+	var regex_res := RegEx.new()
+	var regex_tscn := RegEx.new()
+	regex_tscn.compile(".tscn")
+	regex_res.compile("^(res://)")
+	var regex_res_result = regex_res.search(path)
+	var regex_tscn_result = regex_tscn.search(path)
+	if regex_res_result == null and regex_tscn_result == null:
+		node = get_node(path)
+	elif regex_tscn_result != null:
```
This can be done with [`String.begins_with()`](https://docs.godotengine.org/en/stable/classes/class_string.html#class-string-method-begins-with) and [`String.ends_with()`](https://docs.godotengine.org/en/stable/classes/class_string.html#class-string-method-ends-with), so there's no need for RegEx and it's a bit faster as it is a really easy check.

CollCaz replied:
Didn't know about those functions thx.
Also it's a really cool project of course id want to contribute!

Kubulambula commented on addons/gdshell/commands/default_commands/ls.gd:
```diff
@@ -0,0 +1,73 @@
+		_:
+			path = argv[1]
+			if path[0] != "/":
+				path = "/" + path
```
I would put this in a separate function like `get_ls_root_path()` or something like that.

I like that you added this usability feature, that adds the `/root/` to the path if not present.

CollCaz replied:
Reworked path validation entirely, also made it work by using a dictionary like you suggested, i did add more fields tho.

Kubulambula commented on addons/gdshell/commands/default_commands/ls.gd:
```diff
@@ -0,0 +1,73 @@
+	return {}
+
+
+var root_node: Node
```
I don't really like this stranded variable. It can be done so that it is a parameter for the `_get_all_children()` method like this:
```gdscript
func get_all_children(root_node: Node, node: Node, prefix: String = "", last: bool = true, start : bool = true):
```
The `root_node` would not change at all and it would be recursively passed down and `node` would be the one, that gets processed at each recursion call.

CollCaz replied:
yup, idk why i left it alone like that, even though i made all the other variables parameters so it's not as if i didn't know i could do it this way, just somehow forgot... very silly of me

Kubulambula commented on addons/gdshell/commands/default_commands/ls.gd:
```diff
@@ -0,0 +1,73 @@
+	elif regex_tscn_result != null:
+		node = load(path).instantiate()
+		if node == null:
+			return {"error": 7, "error_string": "File not found"}
```
Why isn't there `ERR_FILE_NOT_FOUND` as below?
```suggestion
			return {"error": ERR_FILE_NOT_FOUND, "error_string": "File not found"}
```

CollCaz commented:
It just occurred to me that i didn't free up the node that gets instantiated when the command is given a .tscn file.
Ill commit a fix as soon as i can

Kubulambula reviewed (changes requested):
It is a good foundation. I just few changes and it would be great apart from the missing `_get_manual()` method, but I would wait with it until the command is finished, so no problem here.

I was messing with it and I am satisfied with the usability, but I think that the average user would be probably interested solely in the scene tree and not the whole tree with the root and singletons.
I suggest that if the user explicitly doesn't type in `/root/` in the input, the root node should be the current scene, which could be obtained with [`get_tree().current_scene`](https://docs.godotengine.org/en/stable/classes/class_scenetree.html#class-scenetree-property-current-scene). (If the user types in `/some/node/path` I would still consider it relative to the `current_scene` path).
There could be a flag like `--absolute` and `-a` for the path to be forced to begin from `/root`.

Also, the command could return the printed three as a [`Dictionary`](https://docs.godotengine.org/en/stable/classes/class_dictionary.html), so that it could be used by some other external command that the user might implement themself. It could also be the way that the command processes the scene internally. It could be just processed into  [`Dictionary`](https://docs.godotengine.org/en/stable/classes/class_dictionary.html) once and then printed out in a separate function (going trough it would be much simpler) and then it could be returned as command result.

The `res://addons/gdshell/demo/demo.tscn` could look like this:
```json
{
"name": "GDShellDemo",
"type": "Control",
"scene_file_path": "res://addons/gdshell/demo/demo.tscn",
"script": true,
"scene_tree_path": "/root/GDShellDemo",
"children": [
	{
		"name": "GDShellIcon",
		"type": "Sprite2D",
		"scene_file_path": "",
		"script": false,
		"scene_tree_path": "/root/GDShellDemo/GDShellIcon",
		"children": [
			{
				"name": "Label",
				"type": "Label",
				"scene_file_path": "",
				"script": false,
				"scene_tree_path": "/root/GDShellDemo/GDShellIcon/Label",
				"children": []
			}
		]
	}
]
}
```

What do you think?
I really appreciate any feedback or ideas and especially the fact, that you want to contribute! ❤️

Kubulambula commented:
> It just occurred to me that i didn't free up the node that gets instantiated when the command is given a .tscn file. Ill commit a fix as soon as i can

Good catch! Didn't even notice that.

CollCaz commented:
Thank you so much for taking the time and pointing out all that!, exactly what i was hoping for when i committed code i wasn't entirely happy with. I knew it needed improvement but didn't know what to do and this kind of feedback is what i was looking for.

 Ill address everything you brought up now that i have time to do so.

CollCaz committed:
Update addons/gdshell/commands/default_commands/ls.gd

Using error code names instead of their actual value

Co-authored-by: Jakub Janšta <kubulambula@gmail.com>

CollCaz committed:
removed the stray variable, ls now starts at current scene by default

CollCaz committed:
ls now starts at current scene, added function that gets a directory of the tree

CollCaz committed:
function that outputs the tree now does so by recursvly going through the dictionary of tree nodes

CollCaz committed:
added the postfix

CollCaz committed:
removed old code

CollCaz committed:
path validation moved to its own function, the ls command will work as long as the path provided exists, it will work if the path is aboslute or relative to the current scene

CollCaz committed:
cleand up the code a bit and fixed some errors

CollCaz committed:
made hover more consitent, removed a test node from GDSHellDemo

CollCaz committed:
minor fixes

CollCaz commented:
All right, here's a list of what i have done.

-  command now works with both absolute and relative paths, no need to specify
- command will work with or without typing the current scene's root node at the start of the path
- command now works by storing all the tree in a dictionary then outputs that as you suggested
- it takes .tscn and .scn files
- hovering on the script icon will reveal the script name
- hovering on node's name will reveal their full tree path
- hovering on the scene icon will reveal the file path of the instanced scene
- it returns the dictionary in the data variable
- the code is a little bit cleaner

Please let me know if i missed anything or if anything needs improvement! and don't be shy with the feedback.
also i edited this in godot 4.1 and forgot to gitignore the godot.project folder so keep that in mind if you decide to merge.

CollCaz committed:
free the instanced scene once done with it

CollCaz committed:
added a space

Kubulambula commented on addons/gdshell/commands/default_commands/ls.gd:
```diff
@@ -0,0 +1,119 @@
+	
+	for i in child_count:
+		#output(get_tree_dict(node.get_child(i), tree_dict, counter))
+		node_dict["children"].append(get_tree_dict(node.get_child(i), root_node, false))
```
This is a bit more readable.

```suggestion
	for child in node.get_children():
		node_dict["children"].append(get_tree_dict(child, root_node, false))
```

Kubulambula commented on addons/gdshell/commands/default_commands/ls.gd:
```diff
@@ -0,0 +1,119 @@
+	var tree_dict = get_tree_dict(starting_node, starting_node)
+	output_tree_dict(tree_dict)
+	
+	data = tree_dict
+	return {}
```
`data` is not returned. I is just an argument of the `_main()` function and you have to manually put it in the `Dictionary`, that gets returned.

```suggestion
	return {
		"error": 0,
		"error_string": "",
		"data": tree_dict,
	}
```

Kubulambula commented on addons/gdshell/commands/default_commands/ls.gd:
```diff
@@ -0,0 +1,119 @@
+extends GDShellCommand
+
+
+signal done_with_instance(instance: Node)
```
This signal is not necessary. See below.
```suggestion
```

Kubulambula commented on addons/gdshell/commands/default_commands/ls.gd:
```diff
@@ -0,0 +1,119 @@
+
+	return node
+
+
+func _on_done_with_instance(node):
+	node.queue_free()
```
If the signal is deleted, this has no use.
```suggestion
```

Kubulambula commented on addons/gdshell/commands/default_commands/ls.gd:
```diff
@@ -0,0 +1,119 @@
+		node_dict["children"].append(get_tree_dict(node.get_child(i), root_node, false))
+
+	if created_an_instance and node.get_parent() == root_node and node.get_index() == root_node.get_child_count() - 1:
+		done_with_instance.emit(root_node)
```
This can be just `queue_free()`. I don't see the need for this to be a signal and it is just more straight forward and readable. If it is needed, it can be very easily added back in.
```suggestion
		root_node.queue_free()
```

Kubulambula commented on addons/gdshell/commands/default_commands/ls.gd:
```diff
@@ -0,0 +1,119 @@
+			current_scene_path = str(current_scene.get_path())
+		
+		path = current_scene_path + "/" + path
+		node = get_node(path)
```
This just does not print an error, if the node is not found. This command handles the error and notifies the user on its own.
```suggestion
		node = get_node_or_null(path)
```

Kubulambula commented on addons/gdshell/commands/default_commands/ls.gd:
```diff
@@ -0,0 +1,119 @@
+			created_an_instance = true
+	
+	elif is_absolute_path:
+		node = get_node("/" + path)
```
I'm not sure, but isn't why is the "/" even needed?
```suggestion
		node = get_node(path)
```

CollCaz replied:
Actually after testing this edited, I think it's better left as is, This is there in case the user doesn't start the path with a /, and if the user does godot will just ignore the extra slash.
For example: ``` ls root/node``` will work
I should have made a comment explaining this in the code, as it does look out of place

Kubulambula replied:
Thanks for the clarification! The comment would be nice for anyone working with it in the future.

Kubulambula commented on addons/gdshell/commands/default_commands/ls.gd:
```diff
@@ -0,0 +1,119 @@
+signal done_with_instance(instance: Node)
+
+
+var created_an_instance:= false
```
If the node is created in `get_starting_node()` it is not added to the `SceneTree` so we can just later check if the `root_node` is outside the tree and then delete it, if it is.
If the node is not created, it must have come from the `SceneTree` and we want to keep it.
```suggestion
```

Kubulambula commented on addons/gdshell/commands/default_commands/ls.gd:
```diff
@@ -0,0 +1,119 @@
+		#output(get_tree_dict(node.get_child(i), tree_dict, counter))
+		node_dict["children"].append(get_tree_dict(node.get_child(i), root_node, false))
+
+	if created_an_instance and node.get_parent() == root_node and node.get_index() == root_node.get_child_count() - 1:
```
As I mentioned above, this is simpler. Alternatively we can check if `get_parent()` returns `null`, but I think, that this is more error resistant, if for some reason we mess with the node in the future.
```suggestion
	if not node.is_inside_tree():
```

Kubulambula commented on addons/gdshell/commands/default_commands/ls.gd:
```diff
@@ -0,0 +1,119 @@
+			return {"error": ERR_FILE_NOT_FOUND, "error_string": "File not found"}
+		else:
+			node = load(path).instantiate()
+			created_an_instance = true
```
This has no use, it he variable is deleted.
```suggestion
```

Kubulambula reviewed (changes requested):
I really like how you've taken the feedback and improved it a lot. I especially like how the bbcode hints on the icons. It's the details, that add up.

Anyway I reviewed some minor things, that just clean up the code a bit, but nothing that would change the way it works.

It would be nice if the PR included only `ls.gd`, but no worries if you don't manage to get the other files out as I can delete them before merging.

Anyway if there is nothing else after this, I will be really satisfied with the result. Good job!

CollCaz committed:
Update addons/gdshell/commands/default_commands/ls.gd

Removed the child_count variable as it is only used once now and no longer needed

Co-authored-by: Jakub Janšta <kubulambula@gmail.com>

CollCaz committed:
Update addons/gdshell/commands/default_commands/ls.gd

Co-authored-by: Jakub Janšta <kubulambula@gmail.com>

CollCaz committed:
Update addons/gdshell/commands/default_commands/ls.gd

Fixed the return of the main funciton

Co-authored-by: Jakub Janšta <kubulambula@gmail.com>

CollCaz committed:
Update addons/gdshell/commands/default_commands/ls.gd

Used a built in signal instead of the custom one, I should have read the docs more...

Co-authored-by: Jakub Janšta <kubulambula@gmail.com>

CollCaz committed:
Update addons/gdshell/commands/default_commands/ls.gd

Co-authored-by: Jakub Janšta <kubulambula@gmail.com>

CollCaz committed:
Update addons/gdshell/commands/default_commands/ls.gd

Now using a built in signal, this  is not needed

Co-authored-by: Jakub Janšta <kubulambula@gmail.com>

CollCaz committed:
Update addons/gdshell/commands/default_commands/ls.gd

Co-authored-by: Jakub Janšta <kubulambula@gmail.com>

CollCaz committed:
Update addons/gdshell/commands/default_commands/ls.gd

Again, not needed after edits

Co-authored-by: Jakub Janšta <kubulambula@gmail.com>

CollCaz committed:
Update addons/gdshell/commands/default_commands/ls.gd

No longer needed

Co-authored-by: Jakub Janšta <kubulambula@gmail.com>

CollCaz committed:
Update addons/gdshell/commands/default_commands/ls.gd

Removed forgotten and no longer needed thing

Co-authored-by: Jakub Janšta <kubulambula@gmail.com>

CollCaz commented:
All right, that was more than i expected...
Thank your for taking the time to comment on everything, I'll take what i learned here with me going forward.
I think i should've gone through the code more times before requesting a review, as I have forgotten to remove some no longer needed code..
Also should have looked into the built in signals in the docs before creating my own.
And code comments could have helped a bit.

Other than that I think it was OK, Hopefully not too annoying.

Thank you for your advice, and please don't hesitate to bring anything else up.

CollCaz committed:
Delete demo.gd

CollCaz committed:
Delete demo.tscn

CollCaz committed:
Delete project.godot

CollCaz committed:
formatting, removed a line

CollCaz committed:
Merge remote-tracking branch 'refs/remotes/origin/main'

CollCaz committed:
t

CollCaz committed:
removed something for testing

Kubulambula commented:
> Other than that I think it was OK, Hopefully not too annoying.

Not annoying at all! I really appreciate any contribution and this one helped a lot and it was better than just OK. Just the fact, that you decided to contribute was more than enough to make me happy.

Anyway, if you don't want to change anything else, I think that now it is ready to be merged.
Thank you for your contribution! ❤️

CollCaz commented:
Thank you! This was a great learning experience thanks to your feedback.
I don't want to add anything else no.

CollCaz committed:
commented on the weird slash thing

Kubulambula committed:
Undo unwanted changes

Kubulambula committed:
Merge branch 'main' of https://github.com/CollCaz/Godot-GDShell into pr/22
